firmware/application/apps/ui_spectrum_painter.hpp

```diff
@@ -0,0 +1,190 @@
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include "ui.hpp"
+#include "ui_widget.hpp"
+
+#include "ui_navigation.hpp"
+#include "ui_tabview.hpp"
+#include "ui_transmitter.hpp"
+#include "baseband_api.hpp"
+
+#include "portapack.hpp"
+#include "message.hpp"
+
+#include "ui_spectrum_painter_image.hpp"
+#include "ui_spectrum_painter_text.hpp"
+
+namespace ui {
+
+class SpectrumPainterView : public View {
+public:
+	SpectrumPainterView(NavigationView& nav);
+	~SpectrumPainterView();
+
+	SpectrumPainterView(const SpectrumPainterView&) = delete;
+	SpectrumPainterView(SpectrumPainterView&&) = delete;
+	SpectrumPainterView& operator=(const SpectrumPainterView&) = delete;
+	SpectrumPainterView& operator=(SpectrumPainterView&&) = delete;
+
+	void focus() override;
+	void paint(Painter& painter) override;
+	
+	std::string title() const override { return "Spec.Painter"; };
+
+private:
+	void on_target_frequency_changed(rf::Frequency f);
+	void set_target_frequency(const rf::Frequency new_value);
+	rf::Frequency target_frequency() const;
+
+	NavigationView& nav_;
+	bool tx_active { false };
+	uint16_t tx_current_line { 0 };
+	uint16_t tx_current_max_lines { 0 };
+	uint16_t tx_current_width { 0 };
+	systime_t tx_timestamp_start { 0 };
+
+	inline uint32_t tx_time_elapsed() {
+		auto now = chTimeNow();
+		return now - tx_timestamp_start;
+	}
+
+	int32_t tx_gain { 47 };
+	bool rf_amp { false };
+
+	SpectrumInputImageView input_image { nav_ };
+	SpectrumInputTextView input_text { nav_ };
+	
+	std::array<View*, 2> input_views { { &input_image, &input_text } };
+
+	TabView tab_view {
+		{ "Image", Color::white(), input_views[0] },
+		{ "Text", Color::white(), input_views[1] }
+	};
+
+	static constexpr int32_t footer_location = 15 * 16 + 8;
+	ProgressBar progressbar {
+		{ 4, footer_location - 16, 240-8, 16 }
+	};
+
+	Labels labels {
+		{ { 10 * 8, footer_location + 1 * 16 }, "GAIN   A:", Color::light_grey() },
+		{ { 1 * 8, footer_location + 2 * 16 }, "BW:      Du:    P:", Color::light_grey() },
+	};
+
+	FrequencyField field_frequency {
+		{ 0 * 8, footer_location + 1 * 16 },
+	};
+	
+	NumberField field_rfgain {
+		{ 14 * 8, footer_location + 1 * 16 },
+		2,
+		{ 0, 47 },
+		1,
+		' '	
+	};
+
+	NumberField field_rfamp {     // previously  I was using "RFAmpField field_rf_amp" but that is general Receiver amp setting.
+		{ 19 * 8, footer_location + 1 * 16 },
+		2,
+		{ 0, 14 },                // this time we will display GUI , 0 or 14 dBs same as Mic App
+		14,
+		' '
+	};
+
+	Checkbox check_loop {
+		{ 21 * 8, footer_location + 1 * 16 },
+		4,
+		"Loop",
+		true
+	};
+
+	ImageButton button_play {
+		{ 28 * 8, footer_location + 1 * 16, 2 * 8, 1 * 16 },
+		&bitmap_play,
+		Color::green(),
+		Color::black()
+	};
+
+	// add control for Bandwidth
+	OptionsField option_bandwidth {
+		{ 4 * 8, footer_location + 2 * 16 },
+		5,
+		{
+			{ "  8k5", 8500 },
+			{ " 11k ", 11000 },
```

Comment:
is there only meant to be one space in front here? (not sure what one is correct, 2 or 1 to reach 5 char total)

Reply:
I'll remove the code duplication

Reply:
Yes , because some BW needs 3 chars but the max 5 char , and it was to blank that field in any option change 
![20230514_152912](https://github.com/eried/portapack-mayhem/assets/86470699/a3e72647-e0be-4992-8bbb-35e61c5dc1a1)

![Screenshot_20230514-152900_Gallery](https://github.com/eried/portapack-mayhem/assets/86470699/b685b060-63e8-491c-8645-2937dd42a268)